README.md
# Multisite Login Logos

[![Build Status](https://travis-ci.org/prontotools/multisite-login-logos.svg?branch=develop)](https://travis-ci.org/prontotools/multisite-login-logos)

Easily change the logo on the WP login screen using WordPress's Customize settings. Choose between the default network logo or any other image you upload.

Developer Guide
---------------

To run, test, and develop the Multisite Login Logos plugin with Docker container, please simply follow these steps:

1. Build the container:

  `$ docker build -t wptest .`
 
2. Test running the PHPUnit on this plugin:

  `$ docker run -it -v $(pwd):/app wptest /bin/bash -c "service mysql start && phpunit"`

Changelog
----------

= 1.0.1 =
- Support SSL if it is enabled.

= 1.0.0 =
- Set custom logo on login page in Theme Customizer.


CHANGELOG.md
Change Log
==========

[1.0.1] 2016-03-29
--------------------
- Support SSL if it is enabled.

[1.0.0] 2016-02-29
--------------------
- Set custom logo on login page in Theme Customizer.


tests/bootstrap.php
<?php

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) $_tests_dir = '/tmp/wordpress-tests-lib';

require_once $_tests_dir . '/includes/functions.php';

function _manually_load_plugin() {
    require dirname( __FILE__ ) . '/../multisite-login-logos.php';
    require '/tmp/wordpress/wp-includes/class-wp-customize-control.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

require $_tests_dir . '/includes/bootstrap.php';


tests/test-multisite-login-logos.php
<?php

require_once( "multisite-login-logos.php" );

class Multisite_Login_Logos_Test extends WP_UnitTestCase {
    private $multisite_login_logos;

    public function setUp() {
        parent::setUp();

        $this->multisite_login_logos = new Multisite_Login_Logos();
    }

    public function test_add_multisite_login_logos_customizer_is_registered_to_customize_register() {
        global $wp_filter;

        $expected = "Multisite_Login_Logos::add_multisite_login_logos_customizer";
        $key      = array_keys( $wp_filter["customize_register"][10] );
        $this->assertEquals( $expected, $key[0] );

        $expected = "add_multisite_login_logos_customizer";
        $function = $wp_filter["customize_register"][10][ $key[0] ]["function"];
        $this->assertEquals( $expected, $function[1] );

        $expected      = 1;
        $accepted_args = $wp_filter["customize_register"][10][ $key[0] ]["accepted_args"];
        $this->assertEquals( $expected, $accepted_args );
    }

    public function test_add_multisite_login_logos_customizer_should_add_login_logo_section() {
        $wp_customize = $this->getMockBuilder( "WP_Customize_Manager" )
            ->setMethods( array( "add_section", "add_setting", "add_control", "get_setting" ) )
            ->getMock();

        $wp_customize->expects( $this->once() )
            ->method( "add_section" )
            ->with(
                "multisite_login_logos_section",
                array(
                    "title"    => "Login Logo",
                    "priority" => 55,
                )
            );

        $this->multisite_login_logos->add_multisite_login_logos_customizer( $wp_customize );
    }

    public function test_add_multisite_login_logos_customizer_should_add_login_logo_settings() {
        $wp_customize = $this->getMockBuilder( "WP_Customize_Manager" )
            ->setMethods( array( "add_section", "add_setting", "add_control", "get_setting" ) )
            ->getMock();

        $wp_customize->expects( $this->at( 1 ) )
            ->method( "add_setting" )
            ->with(
                "multisite_login_logos_settings",
                array(
                    "default" => "1",
                    "type"    => "option",
                )
            );

        $this->multisite_login_logos->add_multisite_login_logos_customizer( $wp_customize );
    }

    public function test_add_multisite_login_logos_customizer_should_add_login_logo_control() {
        $wp_customize = $this->getMockBuilder( "WP_Customize_Manager" )
            ->setMethods( array( "add_section", "add_setting", "add_control", "get_setting" ) )
            ->getMock();

        $wp_customize->expects( $this->at( 2 ) )
            ->method( "add_control" )
            ->with(
                "multisite_login_logos",
                array(
                    "label"    => "Login logo displays",
                    "section"  => "multisite_login_logos_section",
                    "settings" => "multisite_login_logos_settings",
                    "type"     => "radio",
                    "choices"  => array(
                        "1" => "Default network logo",
                        "3" => "Custom logo",
                    ),
                )
            );

        $this->multisite_login_logos->add_multisite_login_logos_customizer( $wp_customize );
    }

    public function test_add_multisite_login_logos_customizer_should_add_login_logo_custom_settings() {
        $wp_customize = $this->getMockBuilder( "WP_Customize_Manager" )
            ->setMethods( array( "add_section", "add_setting", "add_control", "get_setting" ) )
            ->getMock();

        $wp_customize->expects( $this->at( 3 ) )
            ->method( "add_setting" )
            ->with(
                "multisite_login_logos_custom",
                array(
                    "type" => "option",
                )
            );

        $this->multisite_login_logos->add_multisite_login_logos_customizer( $wp_customize );
    }

    public function test_add_multisite_login_logos_customizer_should_add_login_logo_custom_control() {
        $wp_customize = $this->getMockBuilder( "WP_Customize_Manager" )
            ->setMethods( array( "add_section", "add_setting", "add_control", "get_setting" ) )
            ->getMock();

        $image_control_object = new WP_Customize_Image_Control(
            $wp_customize,
            "multisite_login_logos_custom",
            array(
                "label"    => "Custom logo",
                "section"  => "multisite_login_logos_section",
                "settings" => "multisite_login_logos_custom",
            )
        );
        $image_control_object->instance_number = 6;

        $wp_customize->expects( $this->at( 5 ) )
            ->method( "add_control" )
            ->with( $image_control_object );

        $this->multisite_login_logos->add_multisite_login_logos_customizer( $wp_customize );
    }

    public function test_display_multisite_login_logo_is_registered_to_login_head() {
        global $wp_filter;

        $expected = "display_multisite_login_logo";
        $key      = array_keys( $wp_filter["login_head"][99] );
        $this->assertContains( $expected, $key[1] );

        $expected = "display_multisite_login_logo";
        $function = $wp_filter["login_head"][99][ $key[1] ]["function"];
        $this->assertEquals( $expected, $function[1] );

        $expected      = 0;
        $accepted_args = $wp_filter["login_head"][99][ $key[1] ]["accepted_args"];
        $this->assertEquals( $expected, $accepted_args );
    }

    public function test_display_multisite_login_logo_should_show_nothing_if_setting_is_default_logo() {
        update_option( "multisite_login_logos_settings", "1" );

        $expected = "";

        ob_start();
        $this->multisite_login_logos->display_multisite_login_logo();
        $actual = ob_get_clean();

        $this->assertEquals( $expected, $actual );
    }

    public function test_display_multisite_login_logo_should_show_custom_logo_if_setting_is_custom_logo_logo() {
        update_option( "multisite_login_logos_settings", "3" );
        update_option( "multisite_login_logos_custom", "tests/login-logo.png" );

        $expected  = "<style type=\"text/css\">\n";
        $expected .= ".login h1 a {\n";
        $expected .= "background-image: url(tests/login-logo.png);\n";
        $expected .= "width: 100%;\n";
        $expected .= "height: 74px;\n";
        $expected .= "}\n";
        $expected .= "</style>\n";

        ob_start();
        $this->multisite_login_logos->display_multisite_login_logo();
        $actual = ob_get_clean();

        $this->assertEquals( $expected, $actual );
    }

    public function test_display_multisite_login_logo_should_use_https_if_ssl_is_enabled() {
        $_SERVER["HTTPS"] = "on";

        update_option( "multisite_login_logos_settings", "3" );
        update_option( "multisite_login_logos_custom", "http://local.bypronto.dev/wp-content/uploads/2016/03/51323de85a313e11.png" );

        $expected  = "<style type=\"text/css\">\n";
        $expected .= ".login h1 a {\n";
        $expected .= "background-image: url(https://local.bypronto.dev/wp-content/uploads/2016/03/51323de85a313e11.png);\n";
        $expected .= "width: 100%;\n";
        $expected .= "height: 0px;\n";
        $expected .= "}\n";
        $expected .= "</style>\n";

        ob_start();
        $this->multisite_login_logos->display_multisite_login_logo();
        $actual = ob_get_clean();

        $this->assertEquals( $expected, $actual );

        $_SERVER["HTTPS"] = null;
    }
}


class-multisite-login-logos.php
<?php

class Multisite_Login_Logos {
    function __construct() {
        add_action( "customize_register", array( __CLASS__, "add_multisite_login_logos_customizer" ), 10, 1 );
        add_action( "login_head", array( __CLASS__, "display_multisite_login_logo" ), 99, 0 );
    }

    public function add_multisite_login_logos_customizer( $wp_customize ) {
        $wp_customize->add_section( "multisite_login_logos_section", array(
            "title"    => "Login Logo",
            "priority" => 55,
        ) );

        $wp_customize->add_setting( "multisite_login_logos_settings", array(
            "default" => "1",
            "type"    => "option",
        ) );

        $wp_customize->add_control( "multisite_login_logos", array(
            "label"    => "Login logo displays",
            "section"  => "multisite_login_logos_section",
            "settings" => "multisite_login_logos_settings",
            "type"     => "radio",
            "choices"  => array(
                "1" => "Default network logo",
                "3" => "Custom logo",
            ),
        ) );

        $wp_customize->add_setting( "multisite_login_logos_custom", array(
            "type" => "option",
        ) );

        $wp_customize->add_control(
            new WP_Customize_Image_Control( $wp_customize, "multisite_login_logos_custom", array(
                "label"    => "Custom logo",
                "section"  => "multisite_login_logos_section",
                "settings" => "multisite_login_logos_custom",
            ) )
        );
    }

    function display_multisite_login_logo() {
        $multisite_login_logos_settings = get_option( "multisite_login_logos_settings" );
        if ( "3" == $multisite_login_logos_settings ) {
            $logo = get_option( "multisite_login_logos_custom" );
            if ( true == is_ssl() ) {
                $logo = str_replace( "http://", "https://", $logo );
            }

            try {
                $sizes = getimagesize( $logo );
            }
            catch ( Exception $e ) {
                $sizes = array( 0, 0 );
                error_log( "Caught exception: " . $e->getMessage() );
            }

            echo "<style type=\"text/css\">\n";
            echo ".login h1 a {\n";
            echo "background-image: url(" . $logo . ");\n";
            echo "width: 100%;\n";
            echo "height: " . $sizes[1] . "px;\n";
            echo "}\n";
            echo "</style>\n";
        }
    }
}
